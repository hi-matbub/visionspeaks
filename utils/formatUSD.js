/* eslint-disable import/prefer-default-export */
const formatUSD = (number) => {
  // for inconsistent types
  const string = JSON.stringify(parseFloat(number));
  const { length } = string || number;
  const str = `${string.substring(0, length - 2)}.${string.substring(
    length - 2
  )}`;
  return parseFloat(str).toFixed(2);
};

export { formatUSD };
