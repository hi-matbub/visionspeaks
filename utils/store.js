exports.store = {
  name: 'Dog-Gone Goods',
  logo: '/logo.svg',
  hero: '/hero.svg',
  url: '',
  description: 'A dog treat shop',
  about: 'Everyone loves a happy dog!',
  contact: 'youremail@here.com',
  social: {
    facebook: '',
    twitter: '',
    instagram: '',
  },
};
