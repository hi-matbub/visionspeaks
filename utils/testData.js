// used for testing purposes

exports.testData = {
  store: {
    name: 'Dog-Gone Goods',
    logo: '/logo.svg',
    hero: '/hero.svg',
    url: '',
    description: 'A dog treat shop',
    about: 'Everyone loves a happy dog!',
    contact: 'youremail@here.com',
    social: { facebook: '', twitter: '', instagram: '' },
  },
  products: [
    {
      id: 'prod_IRbpyL6fgse3n',
      object: 'product',
      active: true,
      attributes: [],
      created: 1606151601,
      description: 'demo product description',
      images: ['img.jpg'],
      livemode: false,
      metadata: {
        price: 287,
      },
      name: 'Large Demo Product',
      statement_descriptor: null,
      type: 'service',
      unit_label: null,
      updated: 1606153202,
    },
    {
      id: 'prod_IRbouvicOUom16',
      object: 'product',
      active: true,
      attributes: [],
      created: 1606151551,
      description: 'demo product description',
      images: ['img.jpg'],
      livemode: false,
      metadata: {
        price: 287,
      },
      name: 'Demo Product',
      statement_descriptor: null,
      type: 'service',
      unit_label: null,
      updated: 1606153221,
    },
  ],
};
