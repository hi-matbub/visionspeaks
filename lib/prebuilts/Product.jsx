import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import Modal from 'react-bootstrap/Modal';
import ProductPreview from './ProductPreview';
import QtyControl from '../forms/add_to_cart/QtyControl';
import SlimWrapper from '../wrappers/SlimWrapper';
import Loading from '../components/Loading';
import { formatUSD } from '../../utils/formatUSD';

const Product = ({ product, cart, handleUpdatedCartInState, currentBill }) => {
  const [show, setShow] = useState(false);
  const handleClose = () => setShow(false);
  const handleShow = () => setShow(true);
  const [loading, setLoading] = useState(false);

  useEffect(() => {
    if (!product && !cart) setLoading(true);
    else {
      setLoading(false);
    }
  }, [product, cart]);

  return (
    <>
      {loading ? (
        <Loading data={product} />
      ) : (
        <>
          <SlimWrapper className="mb-2 align-items-center justify-content-center">
            <button
              className="btn btn-outline-dark  d-flex w-100"
              type="button"
              onClick={handleShow}
              style={{
                borderRadius: '6px',
                cursor: 'pointer',
              }}
            >
              <img
                src={product.images[0]}
                alt={product.name}
                className="product_img_prc"
              />
              <div className="d-flex w-100 flex-column align-items-end justify-content-center text-right">
                <h5 className="mt-3 mb-2 ellipse">{product.name}</h5>
                <span className="text-muted mb-3">
                  ${formatUSD(product.metadata.price)} USD
                </span>
              </div>
            </button>
          </SlimWrapper>

          <Modal show={show} onHide={handleClose}>
            <button
              type="button"
              onClick={handleClose}
              className="align-self-end mt-4 mr-4 btn btn-outline-dark"
              style={{
                border: 'none',
                background: 'none',
                fontSize: '16px',
              }}
            >
              Go Back
            </button>
            <ProductPreview product={product} />
            <div className="p-4">
              <h2>{product.name}</h2>
              <p>{product.description}</p>
              <p>{product ? `$${formatUSD(product.metadata.price)}` : null}</p>
              <QtyControl
                product={product}
                currentBill={currentBill}
                cart={cart}
                handleUpdatedCartInState={(updatedCart) =>
                  handleUpdatedCartInState(updatedCart)
                }
              />
            </div>
          </Modal>
        </>
      )}
    </>
  );
};

Product.propTypes = {
  cart: PropTypes.shape({}),
  product: PropTypes.shape({
    name: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    metadata: PropTypes.shape({
      price: PropTypes.string.isRequired,
    }).isRequired,
    images: PropTypes.arrayOf(PropTypes.string).isRequired,
  }).isRequired,
  handleUpdatedCartInState: PropTypes.func.isRequired,
  currentBill: PropTypes.number,
};

Product.defaultProps = {
  cart: {},
  currentBill: 0,
};

export default Product;
