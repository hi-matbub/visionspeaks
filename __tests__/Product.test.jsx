import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import Product from '../lib/prebuilts/Product';
import { testData } from '../utils/testData';

const handleUpdatedCartInState = () => {};

it('Product components renders smoke test without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(
    <Product
      product={testData.products[0]} 
      cart={testData.products} 
      handleUpdatedCartInState={handleUpdatedCartInState}
    />
  , div);
  ReactDOM.unmountComponentAtNode(div);
});

it('Product component renders unchanged', () => {
  const tree = renderer.create(
    <Product
      product={testData.products[0]} 
      cart={testData.products} 
      handleUpdatedCartInState={handleUpdatedCartInState}
    />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
