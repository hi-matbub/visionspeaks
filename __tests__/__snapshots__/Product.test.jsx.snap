// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product component renders unchanged 1`] = `
<div
  className="align-items-center justify-content-center mb-2 align-items-center justify-content-center row"
>
  <div
    className="col-lg-8 col-md-9 col-sm-12"
  >
    <button
      className="btn btn-outline-dark  d-flex w-100"
      onClick={[Function]}
      style={
        Object {
          "borderRadius": "6px",
          "cursor": "pointer",
        }
      }
      type="button"
    >
      <img
        alt="Large Demo Product"
        className="product_img_prc"
        src="img.jpg"
      />
      <div
        className="d-flex w-100 flex-column align-items-end justify-content-center text-right"
      >
        <h5
          className="mt-3 mb-2 ellipse"
        >
          Large Demo Product
        </h5>
        <span
          className="text-muted mb-3"
        >
          $
          2.87
           USD
        </span>
      </div>
    </button>
  </div>
</div>
`;
